### PR TITLE
Speed up .mtz export, #924

### LINKED ISSUE #924: dials.export very slow for large numbers of batches
This is a known thing but I could not find a ticket for it - seems that adding batches is linear time on `n_batches` making it quadratic time overall. This is a "feature" of how the MTZ libraries are implemented. 

batch write time

Some hacking around in `iotbx.mtz` can give a factor of ~ 2 speed up (see graph) however this is still `O(n^2)` behaviour.

### PR DESCRIPTION
**Pull request by Anthchirp**
Instead of running O(n) calls to _add_batch(), which runs an O(n) check
and then runs another C library function running another O(n) check for
a total complexity of O(2n²)...

...run one custom function that inserts all batch information at the
same time for O(n) cost in time.

Fixes #924 if used together with the branch of same name in cctbx_project

### CODE REVIEW

File: util/export_mtz.py [hunk 1]
```diff
@@ -125,135 +125,107 @@ def add_dataset(
...
-    o.set_time1(0.0).set_time2(0.0).set_title("Batch {}".format(batch_number))
-    o.set_ndet(1).set_theta(flex.float((0.0, 0.0))).set_lbmflg(0)
-    o.set_alambd(wavelength).set_delamb(0.0).set_delcor(0.0)
-    o.set_divhd(0.0).set_divvd(0.0)
+    i0 = image_range[0]
+    for i in range(n_batches):
+        if experiment.scan:
+            phi_start[i], phi_range[i] = experiment.scan.get_image_oscillation(i + i0)
+        else:
+            phi_start[i], phi_range[i] = 0.0, 0.0
```
Comment (Anthchirp) [inline util/export_mtz.py]: is `flex.float()` guaranteed to be 0-initialised? If so could drop the `else:` clause

Comment (graeme-winter) [reply util/export_mtz.py]: @Anthchirp by default no but we can force it to be... good suggestion. `flex.float(size, init_value)` will do it

**Comment by Anthchirp** [comment]
This seems to produce viable .mtz files which, according to `mtzdmp` are identical to the original ones, however `xxd` shows a few bit differences which I would like to investigate before merging.

**Comment by graeme-winter** [comment]
There are unused areas in MTZ files which are there to "leave space for the future"

Do you get differences at the same place with repeat runs of one version of the code? Possibly this is just uninitialised space being written to file.

**Comment by Anthchirp** [comment]
it's the last bit of the mantissa in phi_range and phi_end, so basically an irrelevant rounding error in the float-flexes. I'm happy with that.

**Pull request merged**